File: drills/python/solutions/nc27_eval_rpn.py
OPERATORS = "+-*/"


def evalRPN(tokens):
    stack = []
    for token in tokens:
        if token in OPERATORS and len(stack) >= 2:
            b = stack.pop()
            a = stack.pop()
            stack.append(apply(token, a, b))
        else:
            stack.append(int(token))
    return stack[-1] if stack else 0


def apply(operator, a, b):
    if operator == "+":
        return a + b
    if operator == "-":
        return a - b
    if operator == "*":
        return a * b
    # // floors, so -3 // 2 is -2; the problem wants truncation towards zero.
    quotient = abs(a) // abs(b)
    return -quotient if (a < 0) != (b < 0) else quotient

File: drills/ts/solutions/nc27_eval_rpn.ts
const OPERATORS = new Set(["+", "-", "*", "/"]);

export function evalRPN(tokens: string[]): number {
  const stack: number[] = [];
  for (const token of tokens) {
    if (OPERATORS.has(token) && stack.length >= 2) {
      const b = stack.pop()!;
      const a = stack.pop()!;
      stack.push(apply(token, a, b));
    } else {
      stack.push(Number(token));
    }
  }
  return stack.length ? stack[stack.length - 1] : 0;
}

function apply(operator: string, a: number, b: number): number {
  if (operator === "+") return a + b;
  if (operator === "-") return a - b;
  if (operator === "*") return a * b;
  // trunc, not floor: the problem wants -3 / 2 to be -1.
  return Math.trunc(a / b);
}
